ts/WoltLabSuite/Core/Component/File/Upload.ts

```diff
@@ -0,0 +1,249 @@
+    return;
+  }
+
+  fileElement.uploadCompleted(result.fileID, result.mimeType, result.link, result.data, result.generateThumbnails);
+
+  if (result.generateThumbnails) {
+    const response = await generateThumbnails(result.fileID);
+    fileElement.setThumbnails(response.unwrap());
+  }
+}
+
+async function getSha256Hash(data: BufferSource): Promise<string> {
+  const buffer = await window.crypto.subtle.digest("SHA-256", data);
+
+  return Array.from(new Uint8Array(buffer))
+    .map((b) => b.toString(16).padStart(2, "0"))
+    .join("");
+}
+
+function clearPreviousErrors(element: WoltlabCoreFileUploadElement): void {
+  element.parentElement?.querySelectorAll(".innerError").forEach((x) => x.remove());
+}
+
+async function resizeImage(element: WoltlabCoreFileUploadElement, file: File): Promise<File> {
+  switch (file.type) {
+    case "image/jpeg":
+    case "image/png":
+    case "image/webp":
+      // Potential candidate for a resize operation.
+      break;
+
+    default:
+      // Not an image or an unsupported file type.
+      return file;
+  }
+
+  const timeout = new Promise<File>((resolve) => {
+    window.setTimeout(() => resolve(file), 10_000);
+  });
+
+  const resizeConfiguration = JSON.parse(element.dataset.resizeConfiguration!) as ResizeConfiguration;
+
+  const resizer = new ImageResizer();
+  const { image, exif } = await resizer.loadFile(file);
+
+  const maxHeight = resizeConfiguration.maxHeight;
+  let maxWidth = resizeConfiguration.maxWidth;
+  if (window.devicePixelRatio >= 2) {
+    const actualWidth = window.screen.width * window.devicePixelRatio;
+    const actualHeight = window.screen.height * window.devicePixelRatio;
+
+    // If the dimensions are equal then this is a screenshot from a HiDPI
+    // device, thus we downscale this to the “natural” dimensions.
+    if (actualWidth === image.width && actualHeight === image.height) {
+      maxWidth = Math.min(maxWidth, window.screen.width);
+    }
+  }
+
+  const canvas = await resizer.resize(image, maxWidth, maxHeight, resizeConfiguration.quality, true, timeout);
+  if (canvas === undefined) {
+    // The resize operation failed, timed out or there was no need to perform
+    // any scaling whatsoever.
+    return file;
+  }
+
+  let fileType: string = resizeConfiguration.fileType;
+  if (fileType === "image/jpeg" || fileType === "image/webp") {
+    fileType = "image/webp";
+  } else {
+    fileType = file.type;
+  }
+
+  const resizedFile = await resizer.saveFile(
+    {
+      exif,
+      image: canvas,
+    },
+    file.name,
+    fileType,
+    resizeConfiguration.quality,
+  );
+
+  return resizedFile;
+}
+
+function validateFile(element: WoltlabCoreFileUploadElement, file: File): boolean {
+  const fileExtensions = (element.dataset.fileExtensions || "*").split(",");
+  for (const fileExtension of fileExtensions) {
+    if (fileExtension === "*") {
+      return true;
+    } else if (file.name.endsWith(fileExtension)) {
+      return true;
+    }
+  }
```

Comment:
This could possibly be cleaner by leveraging `Array.some`:

```
const hasAllowedExtension = fileExtensions.some(…);
if (hasAllowedExtension) {
  return true;
} else {
  innerError(…);
  return false;
}
```

Reply:
Tried it but apart from the nice variable name it just adds more clutter.